### ISSUE TITLE
Memory leak when evaluating model on CPU with dynamic size tensor input.

### ISSUE THREAD

**Issue by seyong92** [pytorch/pytorch#29893]
## To Reproduce

1. Make a simple network.
2. Change a model to eval mode (with torch.no_grad())
3. Evaluate model with dynamic size input.
4. CPU memory increases a lot.
5. I attached a simple toy example that can reproduce this error.
https://github.com/seyong92/pytorch_memory_leak_test/
Run this code with "python3 main.py -ds" (-ds means the input tensor size will be changed)

## Expected behavior

Because it consumes memory unexpectedly, I cannot run my code on the CPU. (It works fine on GPU.)

## Environment

 - PyTorch Version (e.g., 1.0): 1.2.0
 - OS (e.g., Linux): Ubuntu 16.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): python3 main.py -ds
 - Python version: 3.5 / 3.6 (both happen)
 - CUDA/cuDNN version: 10.1, 10.2 / 7
 - GPU models and configuration: GTX 1080 / GTX 1080 Ti / Titan Xp
 - Any other relevant information: It doesn't happen on PyTorch 0.4.1

cc @ezyang @gchanan @zou3519 @jerryzh168 @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh

**Comment by soumith** [comment]
as per https://github.com/pytorch/pytorch/issues/25267#issuecomment-553517072 can you try running with `MKL_DISABLE_FAST_MM=0` env variable

**Comment by seyong92** [comment]
Unfortunately, it still occurs with command "MKL_DISABLE_FAST_MM=0 python3 main.py -ds"

**Comment by luoqing94** [comment]
Did you solve this problem? I also encountered this problem.

**Comment by seyong92** [comment]
@luoqing94 I didn't solve the problem yet.

**Comment by ezyang** [comment]
Please try `LRU_CACHE_CAPACITY=1` instead; this might be a dupe of #27971

**Comment by seyong92** [comment]
Unfortunately, it still happens with the command "LRU_CACHE_CAPACITY=1 python3 main.py -ds" in my machine.

**Comment by ezyang** [comment]
OK, looks like a new one. @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh can you please check?

**Comment by seyong92** [comment]
@ezyang I tried the option you told in PyTorch 1.3.1 with LRU_CACHE_CAPACITY=1 and it does not happen again.

Thanks for your help :)

**Comment by rexxarM** [comment]
>   70.27 MB      79.9%	13628	 	                      -[TorchModule predictImage:width:height:]
  70.27 MB      79.9%	13628	 	                       torch::jit::script::Module::forward(std::__1::vector<c10::IValue, std::__1::allocator<c10::IValue> >)
  70.27 MB      79.9%	13628	 	                        torch::jit::script::Method::operator()(std::__1::vector<c10::IValue, std::__1::allocator<c10::IValue> >, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, c10::IValue, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, c10::IValue> > > const&)
  70.27 MB      79.9%	13628	 	                         torch::jit::Function::operator()(std::__1::vector<c10::IValue, std::__1::allocator<c10::IValue> >, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, c10::IValue, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, c10::IValue> > > const&)
  69.93 MB      79.6%	8501	 	                          torch::jit::GraphExecutorImplBase::run(std::__1::vector<c10::IValue, std::__1::allocator<c10::IValue> >&)
  69.39 MB      78.9%	24	 	                           torch::jit::InterpreterStateImpl::run(std::__1::vector<c10::IValue, std::__1::allocator<c10::IValue> >&)
  69.39 MB      78.9%	24	 	                            torch::jit::InterpreterStateImpl::runImpl(std::__1::vector<c10::IValue, std::__1::allocator<c10::IValue> >&)
  69.39 MB      78.9%	12	 	                             torch::jit::(anonymous namespace)::$_15::__invoke(std::__1::vector<c10::IValue, std::__1::allocator<c10::IValue> >&)
  69.39 MB      78.9%	12	 	                              at::TypeDefault::_convolution(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, c10::ArrayRef<long long>, long long, bool, bool, bool)
  69.39 MB      78.9%	12	 	                               at::native::_convolution(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, c10::ArrayRef<long long>, long long, bool, bool, bool)
  69.39 MB      78.9%	10	 	                                at::TypeDefault::_convolution_nogroup(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, c10::ArrayRef<long long>)
  69.39 MB      78.9%	10	 	                                 at::native::_convolution_nogroup(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, c10::ArrayRef<long long>)
  69.39 MB      78.9%	10	 	                                  at::TypeDefault::_nnpack_spatial_convolution(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>)
  69.39 MB      78.9%	10	 	                                   at::native::_nnpack_spatial_convolution(at::Tensor const&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>)
  68.84 MB      78.3%	1	 	                                    posix_memalign


This problem reproduce again in iOS, but how can I inject `LRU_CACHE_CAPACITY` to it

**Comment by XiaobingSuper** [comment]
@rexxarM, your reported issue is another problem,  **LRU_CACHE_CAPACITY** only works for MKLDNN path, but MKLDNN not support in IOS now.

**Comment by danilopeixoto** [comment]
`os.environ['LRU_CACHE_CAPACITY'] = '1'` works!

### Environment
Python version: 3.6.2
PyTorch version: 1.5.1
Torchvision version: 0.6.1

**Comment by cpatrickalves** [comment]
os.environ['LRU_CACHE_CAPACITY'] = '1' also work for me, THANKS!

**Comment by sabrabano0** [comment]
os.environ['LRU_CACHE_CAPACITY'] = '1' 

**DOSNOT** WORKE FOR ME 


tested pytorch 1.4  and pytorch 1.6 cuda10.1

**Comment by shuiyuejihua** [comment]
os.environ['LRU_CACHE_CAPACITY'] = '1' works!  thanks a lot.

Environment
Python version: 3.6.5
PyTorch version: 1.4.0
cuda 10.1
task: 
multiprocess  on linux cpu ,
 just want to extract features, 
get the out of model resnet50

tried for loop one by one , the memory also up and up

**Comment by jonsneyers** [comment]
For the record (since I recently found this issue searching for a solution to this particular problem), the relevant environment variable is now called `ONEDNN_PRIMITIVE_CACHE_CAPACITY`. See also: https://www.intel.com/content/www/us/en/develop/documentation/onednn-developer-guide-and-reference/top/advanced-topics/primitive-cache.html
